From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: eins0fx <eins0fx@users.noreply.github.com>
Date: Mon, 16 Dec 2024 00:00:00 +0700
Subject: [PATCH] HID: magicmouse: Add T2 trackpad support for kernel 6.18+

This patch adds support for trackpads found on Macs with the T2
Security Chip for kernel 6.18+. The upstream hid-magicmouse.c was
significantly refactored in 6.18, so this is a rewrite of the T2
support based on the original 4004 patch.

Supported devices:
- MacBook Pro 13" 2018-2019 (J140K)
- MacBook Pro 15" 2018-2019 (J132)
- MacBook Pro 16" 2019 (J152F, J680)
- MacBook Air 2018-2020 (J213, J214K, J223, J230K)

This includes:
- T2 trackpad dimension definitions for all models
- Per-device dimension lookup for proper touchpad sizing
- Device ID table entries for all T2 trackpad variants
- Integration with existing multitouch infrastructure

Signed-off-by: eins0fx <eins0fx@users.noreply.github.com>
---
 drivers/hid/hid-magicmouse.c | 231 +++++++++++++++++++++++++++++++++++
 1 file changed, 231 insertions(+)

--- a/drivers/hid/hid-magicmouse.c	2025-12-16 01:12:26.847434977 +0800
+++ b/drivers/hid/hid-magicmouse.c	2025-12-16 01:17:05.419751442 +0800
@@ -112,6 +112,95 @@
 #define TRACKPAD2_RES_Y \
 	((TRACKPAD2_MAX_Y - TRACKPAD2_MIN_Y) / (TRACKPAD2_DIMENSION_Y / 100))
 
+/* T2 MacBook trackpad dimensions */
+#define J140K_TP_DIMENSION_X (float)12100
+#define J140K_TP_MIN_X -5318
+#define J140K_TP_MAX_X 5787
+#define J140K_TP_RES_X \
+	((J140K_TP_MAX_X - J140K_TP_MIN_X) / (J140K_TP_DIMENSION_X / 100))
+#define J140K_TP_DIMENSION_Y (float)8200
+#define J140K_TP_MIN_Y -157
+#define J140K_TP_MAX_Y 7102
+#define J140K_TP_RES_Y \
+	((J140K_TP_MAX_Y - J140K_TP_MIN_Y) / (J140K_TP_DIMENSION_Y / 100))
+
+#define J132_TP_DIMENSION_X (float)13500
+#define J132_TP_MIN_X -6243
+#define J132_TP_MAX_X 6749
+#define J132_TP_RES_X \
+	((J132_TP_MAX_X - J132_TP_MIN_X) / (J132_TP_DIMENSION_X / 100))
+#define J132_TP_DIMENSION_Y (float)8400
+#define J132_TP_MIN_Y -170
+#define J132_TP_MAX_Y 7685
+#define J132_TP_RES_Y \
+	((J132_TP_MAX_Y - J132_TP_MIN_Y) / (J132_TP_DIMENSION_Y / 100))
+
+#define J680_TP_DIMENSION_X (float)16000
+#define J680_TP_MIN_X -7456
+#define J680_TP_MAX_X 7976
+#define J680_TP_RES_X \
+	((J680_TP_MAX_X - J680_TP_MIN_X) / (J680_TP_DIMENSION_X / 100))
+#define J680_TP_DIMENSION_Y (float)10000
+#define J680_TP_MIN_Y -163
+#define J680_TP_MAX_Y 9283
+#define J680_TP_RES_Y \
+	((J680_TP_MAX_Y - J680_TP_MIN_Y) / (J680_TP_DIMENSION_Y / 100))
+
+#define J213_TP_DIMENSION_X (float)13500
+#define J213_TP_MIN_X -6243
+#define J213_TP_MAX_X 6749
+#define J213_TP_RES_X \
+	((J213_TP_MAX_X - J213_TP_MIN_X) / (J213_TP_DIMENSION_X / 100))
+#define J213_TP_DIMENSION_Y (float)8400
+#define J213_TP_MIN_Y -170
+#define J213_TP_MAX_Y 7685
+#define J213_TP_RES_Y \
+	((J213_TP_MAX_Y - J213_TP_MIN_Y) / (J213_TP_DIMENSION_Y / 100))
+
+#define J214K_TP_DIMENSION_X (float)13200
+#define J214K_TP_MIN_X -6046
+#define J214K_TP_MAX_X 6536
+#define J214K_TP_RES_X \
+	((J214K_TP_MAX_X - J214K_TP_MIN_X) / (J214K_TP_DIMENSION_X / 100))
+#define J214K_TP_DIMENSION_Y (float)8200
+#define J214K_TP_MIN_Y -164
+#define J214K_TP_MAX_Y 7439
+#define J214K_TP_RES_Y \
+	((J214K_TP_MAX_Y - J214K_TP_MIN_Y) / (J214K_TP_DIMENSION_Y / 100))
+
+#define J223_TP_DIMENSION_X (float)13200
+#define J223_TP_MIN_X -6046
+#define J223_TP_MAX_X 6536
+#define J223_TP_RES_X \
+	((J223_TP_MAX_X - J223_TP_MIN_X) / (J223_TP_DIMENSION_X / 100))
+#define J223_TP_DIMENSION_Y (float)8200
+#define J223_TP_MIN_Y -164
+#define J223_TP_MAX_Y 7439
+#define J223_TP_RES_Y \
+	((J223_TP_MAX_Y - J223_TP_MIN_Y) / (J223_TP_DIMENSION_Y / 100))
+
+#define J230K_TP_DIMENSION_X (float)12100
+#define J230K_TP_MIN_X -5318
+#define J230K_TP_MAX_X 5787
+#define J230K_TP_RES_X \
+	((J230K_TP_MAX_X - J230K_TP_MIN_X) / (J230K_TP_DIMENSION_X / 100))
+#define J230K_TP_DIMENSION_Y (float)8200
+#define J230K_TP_MIN_Y -157
+#define J230K_TP_MAX_Y 7102
+#define J230K_TP_RES_Y \
+	((J230K_TP_MAX_Y - J230K_TP_MIN_Y) / (J230K_TP_DIMENSION_Y / 100))
+
+#define J152F_TP_DIMENSION_X (float)16000
+#define J152F_TP_MIN_X -7456
+#define J152F_TP_MAX_X 7976
+#define J152F_TP_RES_X \
+	((J152F_TP_MAX_X - J152F_TP_MIN_X) / (J152F_TP_DIMENSION_X / 100))
+#define J152F_TP_DIMENSION_Y (float)10000
+#define J152F_TP_MIN_Y -163
+#define J152F_TP_MAX_Y 9283
+#define J152F_TP_RES_Y \
+	((J152F_TP_MAX_Y - J152F_TP_MIN_Y) / (J152F_TP_DIMENSION_Y / 100))
+
 /**
  * struct magicmouse_sc - Tracks Magic Mouse-specific data.
  * @input: Input device through which we report events.
@@ -151,6 +240,24 @@
 	struct timer_list battery_timer;
 };
 
+static bool magicmouse_is_t2_trackpad(__u32 product)
+{
+	switch (product) {
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F:
+		return true;
+	default:
+		return false;
+	}
+}
+
 static int magicmouse_firm_touch(struct magicmouse_sc *msc)
 {
 	int touch = -1;
@@ -230,8 +337,8 @@
 		state = tdata[7] & TOUCH_STATE_MASK;
 		down = state != TOUCH_STATE_NONE;
 	} else if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
-		   input->id.product ==
-			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+		   input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+		   magicmouse_is_t2_trackpad(input->id.product)) {
 		id = tdata[8] & 0xf;
 		x = (tdata[1] << 27 | tdata[0] << 19) >> 19;
 		y = -((tdata[3] << 30 | tdata[2] << 22 | tdata[1] << 14) >> 19);
@@ -265,7 +372,8 @@
 	 */
 	if (emulate_scroll_wheel &&
 	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 &&
-	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC &&
+	    !magicmouse_is_t2_trackpad(input->id.product)) {
 		unsigned long now = jiffies;
 		int step_x = msc->touches[id].scroll_x - x;
 		int step_y = msc->touches[id].scroll_y - y;
@@ -365,8 +473,8 @@
 		input_report_abs(input, ABS_MT_POSITION_Y, y);
 
 		if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
-		    input->id.product ==
-			    USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC)
+		    input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+		    magicmouse_is_t2_trackpad(input->id.product))
 			input_report_abs(input, ABS_MT_PRESSURE, pressure);
 
 		if (report_undeciphered) {
@@ -377,7 +485,8 @@
 			else if (input->id.product !=
 					 USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 &&
 				 input->id.product !=
-					 USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC)
+					 USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC &&
+				 !magicmouse_is_t2_trackpad(input->id.product))
 				input_event(input, EV_MSC, MSC_RAW, tdata[8]);
 		}
 	}
@@ -505,8 +614,8 @@
 		input_report_rel(input, REL_X, x);
 		input_report_rel(input, REL_Y, y);
 	} else if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
-		   input->id.product ==
-			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+		   input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+		   magicmouse_is_t2_trackpad(input->id.product)) {
 		input_mt_sync_frame(input);
 		input_report_key(input, BTN_MOUSE, clicks & 1);
 	} else { /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
@@ -536,6 +645,56 @@
 	return 0;
 }
 
+static void magicmouse_get_t2_dimensions(__u32 product, int *min_x, int *max_x,
+					 int *min_y, int *max_y,
+					 int *res_x, int *res_y)
+{
+	switch (product) {
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K:
+		*min_x = J140K_TP_MIN_X; *max_x = J140K_TP_MAX_X;
+		*min_y = J140K_TP_MIN_Y; *max_y = J140K_TP_MAX_Y;
+		*res_x = J140K_TP_RES_X; *res_y = J140K_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132:
+		*min_x = J132_TP_MIN_X; *max_x = J132_TP_MAX_X;
+		*min_y = J132_TP_MIN_Y; *max_y = J132_TP_MAX_Y;
+		*res_x = J132_TP_RES_X; *res_y = J132_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT:
+		*min_x = J680_TP_MIN_X; *max_x = J680_TP_MAX_X;
+		*min_y = J680_TP_MIN_Y; *max_y = J680_TP_MAX_Y;
+		*res_x = J680_TP_RES_X; *res_y = J680_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213:
+		*min_x = J213_TP_MIN_X; *max_x = J213_TP_MAX_X;
+		*min_y = J213_TP_MIN_Y; *max_y = J213_TP_MAX_Y;
+		*res_x = J213_TP_RES_X; *res_y = J213_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K:
+		*min_x = J214K_TP_MIN_X; *max_x = J214K_TP_MAX_X;
+		*min_y = J214K_TP_MIN_Y; *max_y = J214K_TP_MAX_Y;
+		*res_x = J214K_TP_RES_X; *res_y = J214K_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223:
+		*min_x = J223_TP_MIN_X; *max_x = J223_TP_MAX_X;
+		*min_y = J223_TP_MIN_Y; *max_y = J223_TP_MAX_Y;
+		*res_x = J223_TP_RES_X; *res_y = J223_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K:
+		*min_x = J230K_TP_MIN_X; *max_x = J230K_TP_MAX_X;
+		*min_y = J230K_TP_MIN_Y; *max_y = J230K_TP_MAX_Y;
+		*res_x = J230K_TP_RES_X; *res_y = J230K_TP_RES_Y;
+		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F:
+	default:
+		*min_x = J152F_TP_MIN_X; *max_x = J152F_TP_MAX_X;
+		*min_y = J152F_TP_MIN_Y; *max_y = J152F_TP_MAX_Y;
+		*res_x = J152F_TP_RES_X; *res_y = J152F_TP_RES_Y;
+		break;
+	}
+}
+
 static int magicmouse_setup_input(struct input_dev *input, struct hid_device *hdev)
 {
 	int error;
@@ -561,8 +720,7 @@
 			__set_bit(REL_HWHEEL_HI_RES, input->relbit);
 		}
 	} else if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
-		   input->id.product ==
-			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+		   input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
 		/* If the trackpad has been connected to a Mac, the name is
 		 * automatically personalized, e.g., "José Expósito's Trackpad".
 		 * When connected through Bluetooth, the personalized name is
@@ -593,6 +751,18 @@
 
 		mt_flags = INPUT_MT_POINTER | INPUT_MT_DROP_UNUSED |
 				INPUT_MT_TRACK;
+	} else if (magicmouse_is_t2_trackpad(input->id.product)) {
+		/* T2 MacBook internal trackpad */
+		__clear_bit(EV_MSC, input->evbit);
+		__clear_bit(BTN_0, input->keybit);
+		__clear_bit(BTN_RIGHT, input->keybit);
+		__clear_bit(BTN_MIDDLE, input->keybit);
+		__set_bit(BTN_MOUSE, input->keybit);
+		__set_bit(INPUT_PROP_BUTTONPAD, input->propbit);
+		__set_bit(BTN_TOOL_FINGER, input->keybit);
+
+		mt_flags = INPUT_MT_POINTER | INPUT_MT_DROP_UNUSED |
+				INPUT_MT_TRACK;
 	} else { /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
 		/* input->keybit is initialized with incorrect button info
 		 * for Magic Trackpad. There really is only one physical
@@ -661,6 +831,24 @@
 		input_abs_set_res(input, ABS_Y, TRACKPAD2_RES_Y);
 		input_abs_set_res(input, ABS_MT_POSITION_X, TRACKPAD2_RES_X);
 		input_abs_set_res(input, ABS_MT_POSITION_Y, TRACKPAD2_RES_Y);
+	} else if (magicmouse_is_t2_trackpad(input->id.product)) {
+		int min_x, max_x, min_y, max_y, res_x, res_y;
+
+		magicmouse_get_t2_dimensions(input->id.product, &min_x, &max_x,
+					     &min_y, &max_y, &res_x, &res_y);
+
+		input_set_abs_params(input, ABS_MT_PRESSURE, 0, 253, 0, 0);
+		input_set_abs_params(input, ABS_PRESSURE, 0, 253, 0, 0);
+		input_set_abs_params(input, ABS_MT_ORIENTATION, -3, 4, 0, 0);
+		input_set_abs_params(input, ABS_X, min_x, max_x, 0, 0);
+		input_set_abs_params(input, ABS_Y, min_y, max_y, 0, 0);
+		input_set_abs_params(input, ABS_MT_POSITION_X, min_x, max_x, 0, 0);
+		input_set_abs_params(input, ABS_MT_POSITION_Y, min_y, max_y, 0, 0);
+
+		input_abs_set_res(input, ABS_X, res_x);
+		input_abs_set_res(input, ABS_Y, res_y);
+		input_abs_set_res(input, ABS_MT_POSITION_X, res_x);
+		input_abs_set_res(input, ABS_MT_POSITION_Y, res_y);
 	} else { /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
 		input_set_abs_params(input, ABS_MT_ORIENTATION, -31, 32, 1, 0);
 		input_set_abs_params(input, ABS_X, TRACKPAD_MIN_X,
@@ -684,7 +872,8 @@
 
 	if (report_undeciphered &&
 	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 &&
-	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
+	    input->id.product != USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC &&
+	    !magicmouse_is_t2_trackpad(input->id.product)) {
 		__set_bit(EV_MSC, input->evbit);
 		__set_bit(MSC_RAW, input->mscbit);
 	}
@@ -710,8 +899,8 @@
 	/* Magic Trackpad does not give relative data after switching to MT */
 	if ((hi->input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD ||
 	     hi->input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
-	     hi->input->id.product ==
-		     USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) &&
+	     hi->input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC ||
+	     magicmouse_is_t2_trackpad(hi->input->id.product)) &&
 	    field->flags & HID_MAIN_ITEM_RELATIVE)
 		return -1;
 
@@ -760,6 +949,18 @@
 			feature = feature_mt_trackpad2_usb;
 		}
 		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F:
+		feature_size = sizeof(feature_mt_trackpad2_usb);
+		feature = feature_mt_trackpad2_usb;
+		break;
 	case USB_DEVICE_ID_APPLE_MAGICMOUSE2:
 	case USB_DEVICE_ID_APPLE_MAGICMOUSE2_USBC:
 		feature_size = sizeof(feature_mt_mouse2);
@@ -916,6 +1117,18 @@
 				TRACKPAD2_USB_REPORT_ID, 0);
 		}
 		break;
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K:
+	case USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F:
+		report = hid_register_report(hdev, HID_INPUT_REPORT,
+			TRACKPAD2_USB_REPORT_ID, 0);
+		break;
 	default: /* USB_DEVICE_ID_APPLE_MAGICTRACKPAD */
 		report = hid_register_report(hdev, HID_INPUT_REPORT,
 			TRACKPAD_REPORT_ID, 0);
@@ -1023,6 +1236,24 @@
 		USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC), .driver_data = 0 },
 	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
 		USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J140K), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J132), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J680_ALT), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J213), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J214K), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J223), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J230K), .driver_data = 0 },
+	{ HID_USB_DEVICE(USB_VENDOR_ID_APPLE,
+		USB_DEVICE_ID_APPLE_WELLSPRINGT2_J152F), .driver_data = 0 },
 	{ }
 };
 MODULE_DEVICE_TABLE(hid, magic_mice);
-- 
2.47.0
